### PR TITLE
Allow mutable state in maybe_scalar_fun wrapper

### PR DESCRIPTION
**Pull request by yannra**
Hi all,
In general, mutable states of the models are supported by netket. If these are used, the model apply_fun will generally return a tuple where the second element contains the state indicated as mutable. The recent commit e27386c1483c531e788d977781caf85635b9074f does however explicitly assume that only the model value is returned by the model apply_fun (which breaks if a tuple is returned). This PR fixes this by checking if the return type is a tuple or not.

Hope this makes sense.

Best,
Yannic

### CODE REVIEW

**Comment by PhilipVinc** [comment]
Hi @yannra and welcome and thank you!

When I wrote most of NetKet 3 I tried to lay the foundations in order to support mutable state, but that is relatively spotty and often is not fully implemented. The main problem arises from the fact that it's hard to know how to combine mutable states when computing gradients.

But if we can fix it and make netket work with mutable state, that would be great!

Indeed when I added that commit I forgot about mutable state. We would need at least a test that exercises this functionality. Would you be willing to add it?

BTW, could you also rebase your commit on current master? I see that there is a few changes from Netket's past commits, and it's harder to review your PR that way.

**Comment by PhilipVinc** [comment]
Could you also add a test [in test/utils/test_utils.py](https://github.com/netket/netket/blob/e27386c1483c531e788d977781caf85635b9074f/test/utils/test_utils.py#L93) exercising this new functionality?

**Comment by yannra** [comment]
Sure, no problem, I have now added the simple test case to the PR.

Re the full integration of mutable states into the code and the test cases, I'm not entirely sure what the original intention was to allow for mutable states in the variational state interface. If you could give a little bit more background what it could/should be used for, I can definitely look into fully integrating that into the code/test cases.

Maybe to give a bit of background how we are using this and why I created this PR in the first place:
We are currently using the mutable argument of the apply function to store some intermediate values in order to speed up the calculation of the local energy in our netket calculations (which can quite often be done as the connected configurations are related to the root sample via a constant number of local updates). Our implementation does however not really use the "mutable" interface of the variational state interface but it is directly hacked into the local energy evaluation. We'll obviously share and/or integrate our netket add-ons once these are in a presentable and usable state.

File: netket/utils/jax.py [hunk 1]
```diff
@@ -74,7 +74,12 @@ def wrap_to_support_scalar(fun):
     def maybe_scalar_fun(apply_fun, pars, x, *args, **kwargs):
         xb = jnp.atleast_2d(x)
         res = apply_fun(pars, xb, *args, **kwargs)
-        res = res.reshape(()) if x.ndim == 1 else res
+        if isinstance(res, tuple):
```
Comment (PhilipVinc) [inline netket/utils/jax.py:77]:
```suggestion
        # support models with mutable state
        if isinstance(res, tuple):
```

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
Anyhow, thanks for the contribution!
